### PR TITLE
Pandas extension dtype for image

### PR DESCRIPTION
**Pull request by changhiskhan**
We have the Image arrow extension type but it breaks to_pandas conversion.
Here we implement a pandas extension dtype for ImageArray with ImageUri's.
It is backed by a string ndarray. 

Note that this requires copying the strings
which impacts performance a lot. A later optimization could be to directly
back this with arrow array.

References:

https://pandas.pydata.org/docs/development/extending.html#extension-types

https://pandas.pydata.org/docs/development/extending.html#testing-extension-arrays

### CODE REVIEW

File: python/lance/types/pandas/base.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    def __getitem__(self, *args):
+        result = operator.getitem(self.data, *args)
+        if isinstance(result, tuple):
+            return self._box_scalar(result)
+        elif result.ndim == 0:
+            return self._box_scalar(result.item())
+        else:
+            return type(self)(result)
+
+    def setitem(self, indexer, value):
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/lance/types/pandas/base.py:39]: no i should remove this file actually. nice catch

File: python/test_mac.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+
+pushd ..
+CIBW_TEST_SKIP=* CIBW_BUILD=cp310* python -m cibuildwheel --platform macos python && \
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/test_mac.sh:4]: i'll make it generic. Saves so much time i'd like to have it available by default

File: python/lance/types/pandas/image.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+        is_scalar = self._is_scalar_value(value)
+        if is_scalar:
+            value = _convert_value(value)
+        return super()._where(mask, value)
+
+    @property
+    def nbytes(self):
+        return self._ndarray.nbytes
+
+    # -------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/lance/types/pandas/image.py:116]: done

File: python/lance/types/pandas/image.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/lance/types/pandas/image.py:14]: done

File: python/lance/tests/types/pandas/test_image.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        ser = pd.Series(data)
+        self._compare_other(ser, data, comparison_op, "abc")
+
+    def test_compare_array(self, data, comparison_op):
+        ser = pd.Series(data)
+        other = pd.Series([str(data[0])] * len(data))
+        self._compare_other(ser, data, comparison_op, other)
+
+
+class TestParsing(base.BaseParsingTests):
```
Comment (eddyxu) [inline python/lance/tests/types/pandas/test_image.py:159]: What are those empty class used for?

**Review by eddyxu — APPROVED** [review]
One question. no block tho.

File: python/lance/tests/types/pandas/test_image.py (hunk 5 above)
Comment (changhiskhan) [reply python/lance/tests/types/pandas/test_image.py:159]: Basically these are test suites for every pandas ExtensionArray. When you make a new one, you just override the fixtures and override/skip tests that are specific to your ExtensionArray

**Pull request merged**